snuba/subscriptions/scheduler_processing_strategy.py

```diff
@@ -257,3 +278,167 @@ def terminate(self) -> None:
+            Tuple[Message[CommittableTick], Deque[Future[Message[KafkaPayload]]]]
+        ] = deque()
+
+    def append(
+        self,
+        tick_message: Message[CommittableTick],
+        futures: Deque[Future[Message[KafkaPayload]]],
+    ) -> None:
+        if len(futures) > 0:
+            self.__queues.append((tick_message, futures))
+
+    def peek(self) -> Optional[TickSubscription]:
+        if self.__queues:
+            tick, futures = self.__queues[0]
+            return TickSubscription(tick, futures[0], len(futures) == 1)
+        return None
+
+    def popleft(self) -> TickSubscription:
+        if self.__queues:
+            tick_message, futures = self.__queues[0]
+            subscription_future = futures.popleft()
+
+            is_empty = len(futures) == 0
+
+            if is_empty:
+                self.__queues.popleft()
+
+            should_commit = is_empty and tick_message.payload.should_commit == True
+
+            return TickSubscription(tick_message, subscription_future, should_commit)
+
+        raise IndexError()
+
+    def __len__(self) -> int:
+        return sum([len(futures) for (_tick, futures) in self.__queues])
+
+
+class ProduceScheduledSubscriptionMessage(ProcessingStrategy[CommittableTick]):
+    """"
+    This strategy is responsible for producing a message for all of the subscriptions
+    scheduled for a given tick.
+
+    The subscriptions to be scheduled are those that:
+    - Have the same partition index as the tick received, and
+    - Are scheduled to be run within the time interval defined by the tick
+    (taking the jitter into account)
+
+    The subscriptions to be scheduled are encoded and produced to the
+    scheduled topic to be picked up by the subscription executor which
+    will run the actual query later.
+
+    When a tick is submitted to this strategy, we produce all of it's
+    subscriptions and place it in a queue without waiting to see if they
+    were succesfully produced or not.
+
+    On each call to poll(), we remove completed futures from the queue.
+    If all scheduled subscription messages for that tick were succesfully
+    produced and the tick indicates it should be committed, we commit offsets.
+    """
+
+    def __init__(
+        self,
+        entity_key: EntityKey,
+        schedulers: Mapping[int, Scheduler[Subscription]],
+        producer: Producer[KafkaPayload],
+        scheduled_topic_spec: KafkaTopicSpec,
+        commit: Callable[[Mapping[Partition, Position]], None],
+    ) -> None:
+        self.__schedulers = schedulers
+        self.__encoder = SubscriptionScheduledTaskEncoder(entity_key)
+        self.__producer = producer
+        self.__scheduled_topic = Topic(scheduled_topic_spec.topic_name)
+        self.__commit = commit
+        self.__closed = False
+
+        # Stores each tick with it's futures
+        self.__queue = ScheduledSubscriptionQueue()
+
+        # Not a hard max
+        self.__max_buffer_size = 10000
```

Comment:
Setting ?

Reply:
I'm not convinced we actually need to make this configurable- I think we can probably get away with hardcoding a reasonable value. I can't think of a reason why this would need to differ for different datasets right now. I think we can move it to settings later if there is a strong justification for being able to override it.